Cubical/HITs/SetTruncation/Properties.agda

```diff
@@ -75,6 +76,121 @@ elim3 Bset g = elim2 (λ _ _ → isSetΠ (λ _ → Bset _ _ _))
+                              {B = λ p → PathP (λ i → P (p i)) (η* a) (η* b)}
+                              (λ _ → isOfHLevelPathP 1 (Pprop _) _ _) _ _ (δ a b p)
+
+ -- The main trick: eliminating into hsets is easy
+ -- i.e. H has the universal property of set truncation...
+ module HelimSet {P : H → Type ℓ''} (Pset : ∀ h → isSet (P h))
+                 (η* : ∀ a → P (η a)) where
+
+  fun : (h : H) → P h
+  fun = Helim.fun (λ _ → isSet→isGroupoid (Pset _)) η* ε*
+                   λ a b p → isOfHLevel→isOfHLevelDep 1
+                             {B = λ p → PathP (λ i → P (p i)) (η* a) (η* b)}
+                             (λ _ → isOfHLevelPathP' 1 (Pset _) _ _) _ _ (δ a b p)
+   where
+   ε* : (a b : A) (∣p∣ : ∥ a ≡ b ∥) → PathP (λ i → P (ε a b ∣p∣ i)) (η* a) (η* b)
+   ε* a b = pElim (λ _ → isOfHLevelPathP' 1 (Pset _) (η* a) (η* b))
+                   λ p → subst (λ x → PathP (λ i → P (x i)) (η* a) (η* b))
+                               (sym (δ a b p)) (cong η* p)
+
+
+ -- our desired function will split through H:
+ -- ∥ A ∥₂ → H → B
+ -- we first define the rhs function
+ f₁ : H → B
+ f₁ = Hrec.fun Bgpd f εᶠ λ _ _ _ → refl
+  where
+  εᶠ : (a b : A) → ∥ a ≡ b ∥ → f a ≡ f b
+  εᶠ a b = rec→Set (Bgpd _ _) (cong f) λ p q → congFConst a b p q
+  -- this is the inductive step,
+  -- we use that maps ∥ A ∥ → B for an hset B
+  -- correspond to 2-Constant maps A → B
+
+ -- Now we need to prove that H is a set.
+ -- From that we immediately get the desired result...
+ -- upstream lemma?:
+ localHedbergLemma : {X : Type ℓ''} (P : X → Type ℓ'')
+                   → (∀ x → isProp (P x))
+                   → ((x y : X) → P x → P y → x ≡ y)
+                  --------------------------------------------------
+                   → (x : X) → P x → (y : X) → isProp (x ≡ y)
+ localHedbergLemma {X = X} P Pprop P→≡ x px y = isPropRetract
+                   (λ p → subst P p px) (λ py → sym (P→≡ x x px px) ∙ P→≡ x y px py)
+                   isRetract (Pprop y)
+  where
+  isRetract : (p : x ≡ y) → (sym (P→≡ x x px px)) ∙ P→≡ x y px (subst P p px) ≡ p
+  isRetract p = J (λ y' p' → (sym (P→≡ x x px px)) ∙ P→≡ x y' px (subst P p' px) ≡ p')
+                  (subst (λ px' → sym (P→≡ x x px px) ∙ P→≡ x x px px' ≡ refl)
+                  (sym (substRefl {B = P} px)) (lCancel (P→≡ x x px px))) p
+
+ Hset : isSet H
+ Hset = HelimProp.fun (λ _ → isPropΠ λ _ → isPropIsProp) baseCaseLeft
+  where
+  baseCaseLeft : (a₀ : A) (y : H) → isProp (η a₀ ≡ y)
+  baseCaseLeft a₀ = localHedbergLemma (λ x → Q x .fst) (λ x → Q x .snd) Q→≡ _ ∣ refl ∣
+   where
+   Q : H → hProp ℓ
+   Q = HelimSet.fun (λ _ → isSetHProp) λ b → ∥ a₀ ≡ b ∥ , propTruncIsProp
+   -- Q (η b) = ∥ a ≡ b ∥
+
+   Q→≡ : (x y : H) → Q x .fst → Q y .fst → x ≡ y
+   Q→≡ =     HelimSet.fun (λ _ → isSetΠ3 λ _ _ _ → gtrunc _ _)
+       λ a → HelimSet.fun (λ _ → isSetΠ2 λ _ _ → gtrunc _ _)
+       λ b p q → sym (ε a₀ a p) ∙ ε a₀ b q
+
+ f₂ : ∥ A ∥₂ → H
+ f₂ = rec Hset η
+
+ fun : ∥ A ∥₂ → B
+ fun = f₁ ∘ f₂
```

Comment:
Maybe restate this at the top-level to make it easier to figure out what the statement really is?

Reply:
should be clear now, I think